File: ch4/first_tree/python/Tree.py
from Node import *

class Tree:
    def __init__(self, d):
        self.head = Node(d)

    def insert(self, d):
        newNode = Node(d)
        tmpHead = self.head
        while(tmpHead != None):
            if(newNode.data < tmpHead.data):
                if(tmpHead.left == None):
                    tmpHead.left = newNode
                    break
                else:
                    tmpHead = tmpHead.left
            elif(newNode.data > tmpHead.data):
                if(tmpHead.right == None):
                    tmpHead.right = newNode
                    break
                tmpHead = tmpHead.right
        tmpHead = newNode

    def preList(self):
        return self.preListR(self.head)

    def preListR(self, n):
        if(n.left == ):


    def p(self):
        tmpHead = self.head
        self.pR(tmpHead)

    def pR(self, n):
        if(n != None):
            print(n.data)
        if(n.left != None):
            self.pR(n.left)
        if(n.right != None):
            self.pR(n.right)

File: ch4/first_tree/java/Tree.java
// min heap
import java.lang.Math;

public class Tree{
    Node head;
    Tree(int i){
        head = new Node(i);
    }
    
    // Get Minimum item //
    public int min(){
        return head.data;
    }

    public void insert(int d){
        Node newNode = new Node(d);
        Node start = head;
        while(start.right != null){
            start = start.right;
        }
        start.right = head;
    }

    // Inorder Traversal //
    public void inorder(){
        inorderR(head);
        System.out.println("");
    }
    public void inorderR(Node start){
        if(start.left != null){
            inorderR(start.left);
        }
        System.out.print(start.data);
        System.out.print(" ");
        if(start.right != null){
            inorderR(start.right);
        }
    }

    // Preorder Traversal //
    public void preorder(){
        preorderR(head);
        System.out.println("");
    }
    public void preorderR(Node start){
        System.out.print(start.data);
        System.out.print(" ");
        if(start.left != null){
            preorderR(start.left);
        }
        if(start.right != null){
            preorderR(start.right);
        }
    }

    // Postorder Traversal //
    public void postorder(){
        postorderR(head);
        System.out.println("");
    }
    public void postorderR(Node start){
        if(start.left != null){
            postorderR(start.left);
        }
        if(start.right != null){
            postorderR(start.right);
        }
        System.out.print(start.data);
        System.out.print(" ");
    }
}